Server.cpp

```diff
@@ -1,82 +1,327 @@
+		throw (std::runtime_error(std::string("listen: ") + std::strerror(errno)));
+
+	log("Socket %d is listening on port %d.", this->_socket, this->_port);
+
+	set_signal();
+}
+
+/**
+ * @brief Stop the server
+ *
+ * Stop the server by disconnecting every client and closing associated sockets
+ * Should be called just before server desctructor / End of program
+ */
+void		Server::stopServer( void ) {
+	if (this->_pollFd.size() > 1) {
+		log("Disconecting every client...");
+		unsigned long int i = this->_pollFd.size() - 1;
+		while (i > 0)
+			disconnectClient(i);
+	}
+	log("%sStopping server%s", WHT, RESET);
+	if (this->_socket != -1) {
+		log("Closing socket %d.", this->_socket);
+		close(this->_socket);
+	}
+	if (this->_serverInfo != NULL) {
+		log("Freeing server informations");
+		freeaddrinfo(this->_serverInfo);
+	}
+	log("Server stopped!");
+}
+
+/**
+ * @brief Accept a new client connecting to the server
+ *
+ * Accept new client after a POLLIN event have been caught on the Server socket
+ */
+void	Server::acceptNewClient( void ) {
+	int					client_socket;
+	struct sockaddr_in	client_addr;
+	int					addr_len = sizeof(client_addr);
+
+		client_socket = accept(this->_socket, (struct sockaddr*)&client_addr,
+			reinterpret_cast<socklen_t *>(&addr_len));
+		if (client_socket < 0) //!< accept error
+			return (err_log("%sInternal error%s: %saccept%s: %s",
+					RED, RESET, MAG, RESET, std::strerror(errno)));
+		success_log("accepted client on: %s", inet_ntoa(client_addr.sin_addr));
+		this->pollPushBack(client_socket, POLLIN); //!< add client to poll
+}
+
+/**
+ * @brief Send msg to all client except the one specified by fd
+ *
+ * This function is needed to broadcast the irc message when needed
+ *
+ * @param buffer The buffer containing the message to send
+ * @param len The message len
+ * @param fd The client exluded from receiving the message (the sender)
+ */
+void	Server::msgToAllExceptOne(const char *buffer, size_t len, int fd) {
+	int	send_result;
+
+	for (unsigned long int i = 1; i < this->_pollFd.size(); i++) {
+		if (this->_pollFd[i].fd == fd)
+			continue ;
+		log("Sending back msg received on socket %d to %d", fd, this->_pollFd[i].fd);
+		send_result = send(this->_pollFd[i].fd, buffer, len, MSG_DONTWAIT);
+		if (send_result == -1) {
+			err_log("Msg could not be sent on socket %d", this->_pollFd[i].fd);
+			continue ; }
+		if (len != send_result)
+			war_log("Msg could not be sent entirely on socket %d", this->_pollFd[i].fd);
+	}
+}
+
+/**
+ * @brief Receive client message after POLLIN event
+ *
+ * This function read the Msg send byy the client after POLLIN event
+ * IT SHOULD BE UPDATED ACCORDINGLY
+ *
+ * @param i The pollFd index
+ */
+void	Server::receiveMsg( long unsigned int& i ) {
+	char	buffer[512]; //<! 512 = max irc message size
+	
+	int buffer_size = recv(this->_pollFd[i].fd, buffer, 512 - 1, MSG_DONTWAIT);
+
+	if (buffer_size > 0) { //!< Received msg
+		buffer[buffer_size - 1] = '\0';
```

Comment:
quand je test ton serveur, parfois le message envoye est recu par un autre client sans '\n', et du coup le message de client 1 est attache au message du client 2
exemple (how are you est un msg envoye par le client 2 au client 1 qui la recu sans new line)
![Screenshot from 2024-11-15 16-43-33](https://github.com/user-attachments/assets/b9899a11-fcac-46a8-9627-225c28ce05dc)

Reply:
Oui normalement c'est buffer[buffer_size] = 0. Pour mes tests je voulais pas le '\n' a la fin je trouvais ca pas beau